### PR TITLE
[MRG] LRScheduler batch option

### LINKED ISSUE #610: Cosine annealing called every epoch instead of every batch
Cosine annealing LR should update the learning rate every batch according to: https://arxiv.org/pdf/1608.03983.pdf

It does not, seeing as the method `skorch.LRScheduler.on_batch_end` is: 
```python
    def on_batch_end(self, net, training, **kwargs):
        if (
                training and
                hasattr(self.lr_scheduler_, 'batch_step') and
                callable(self.lr_scheduler_.batch_step)
        ):
            self.lr_scheduler_.batch_step(self.batch_idx_)

        if TorchCyclicLR and isinstance(self.lr_scheduler_, TorchCyclicLR):
            self.lr_scheduler_.step(self.batch_idx_)

        if training:
            self.batch_idx_ += 1
```
`torch.optim.lr_scheduler.CosineAnnealingLR` is neither `TorchCyclicLR` nor has a `batch_step` call (On that note, I haven't found a single LRScheduler that does have a `batch_step` method).

And on a somewhat unrelated note, it would be ideal for me to log somehow the LR per batch (or epoch, depending on the scheduler). Would it be wise to add a `net.record(...)` / `net.record_batch` call to `skorch.LRScheduler`?

### PR DESCRIPTION
**Pull request by guybuk**
Closes #610 

I could use some assistance with schedulers that update once an epoch:

`optim._LRScheduler` initializes with `self.last_epoch=0`. This means that if we're training for `0,1,2,3,4` epochs, the respective values for `last_epoch` at the end of every epoch (where we update LR) will be `1,2,3,4,5`. This requires changes I made in the tests:

`assert lr_policy.lr_scheduler_.last_epoch == max_epochs-1` => `assert lr_policy.lr_scheduler_.last_epoch == max_epochs`

The reason this seems to have worked until now is in fact because of what I think is a bug:

It seems like until now, inside `on_epoch_end`, `epoch=len(net.history)-1=>epoch=0`. This means  that when `self.lr_scheduler_.step(epoch)` is called for the first time, the scheduler does not actually take a step because `self.last_step==epoch==0`, and `last_epoch` would stay at 0 for one extra epoch, which is why the test above passes (the one with `max_epochs - 1`).

I changed this code so that `epoch=len(history)` so the step would make sense, and replaced the order of `history.record` with `lr_scheduler_.step` so that from now on `history.record` will record **the LR with which the step was performed rather than the LR with which the next step will be performed** (which was this way until now).

Finally, I feel like as far as I can tell, all is in order. Except for `ReduceLROnPlateau`, which I'm struggling to understand how it works in general, and how the unique code for it inside `on_epoch_end` works.

Thanks!

### CODE REVIEW

**Comment by BenjaminBossan** [comment]
Thanks again for working on this @guybuk. I'll take a closer look when I have more time. 

@thomasjpfan is the export for LR schedulers, hopefully he has time to answer your questions.

File: skorch/tests/callbacks/test_lr_scheduler.py [hunk 1]
```diff
@@ -94,10 +94,10 @@ def test_lr_callback_steps_correctly(
         )
         net.fit(X, y)
         # pylint: disable=protected-access
-        assert lr_policy.lr_scheduler_.last_epoch == max_epochs - 1
+        assert lr_policy.lr_scheduler_.last_epoch == max_epochs
 
     @pytest.mark.parametrize('policy, kwargs', [
-        (TorchCyclicLR, {'base_lr': 1e-3, 'max_lr': 6e-3}),
+        (TorchCyclicLR, {'base_lr': 1e-3, 'max_lr': 6e-3, 'step_every': 'batch'}),
```
Comment (thomasjpfan) [inline skorch/tests/callbacks/test_lr_scheduler.py:100]: This would not be backwards compatible with `TorchCyclicLR` with default arguments. As in `TorchCyclicLR` would step every epoch if a user does not pass in `step_every='batch'`.

File: skorch/callbacks/lr_scheduler.py [hunk 2]
```diff
@@ -57,6 +57,10 @@ class LRScheduler(Callback):
       Pass ``None`` to disable placing events in history.
       **Note:** This feature works only for pytorch version >=1.4
 
+    step_every: str, (default='epoch'_
```
Comment (thomasjpfan) [inline skorch/callbacks/lr_scheduler.py:60]:
```suggestion
    step_every: str, (default='epoch')
```

**Comment by BenjaminBossan** [comment]
> For skorch, since we are support torch > 1.1, this change should be okay.

We actually do support 1.1 to 1.4. However, I was thinking of adding 1.5, maybe we should drop 1.1 then? My impression is that since PyTorch introduces so few breaking changes, most people use the more recent versions, but I have no data to back that up.

**Comment by thomasjpfan** [comment]
Sorry I mistyped. I meant to say that we support `pytorch >= 1.1` (with the equality), we can depend on the new behavior. So we do not need to drop `1.1`.

File: skorch/tests/callbacks/test_lr_scheduler.py (hunk 1 above)
Comment (guybuk) [reply skorch/tests/callbacks/test_lr_scheduler.py:100]: So the implementation is fine and we just need to restore the ` if TorchCyclicLR and isinstance(self.lr_scheduler_, TorchCyclicLR):...` segment?

Comment (thomasjpfan) [reply skorch/tests/callbacks/test_lr_scheduler.py:100]: Now that `step_every` is a parameter to `LRScheduler`, if we keep the `isinstance(self.lr_scheduler_, TorchCyclicLR)` piece, we will not be strictly following the `step_every` parameter.

We can have a `step_every='auto'` option that special cases the `CosineAnnealingLR` and `TorchCyclicLR`. Specifically, if the class is `CosineAnnealingLR` or `TorchCyclicLR` we use batch steps, otherwise we use epoch step. We will still allow the `batch` and `epoch` options for custom lr schdulers.

Comment (guybuk) [reply skorch/tests/callbacks/test_lr_scheduler.py:100]: That would mean introducing the entire concept of an 'auto' option ,and having to be backwards compatible for it in the future. However all it'll do is special case cyclic (for the record, cosine annealing isn't hard coded to work with batches, only cyclic). 

This basically means that if we don't want 'auto'  to just be a cover up for the cyclic case and instead be meaningful, we will have to add reasonable recommendations for every scheduler we introduce (which will be the default auto options). That will most likely involve having to read the paper from which it was introduced, or it's documentation.

Comment (BenjaminBossan) [reply skorch/tests/callbacks/test_lr_scheduler.py:100]: This is a bit of a conundrum. Do you know why some schedulers prefer batch and other epoch? Is it some derived property or more arbitrary?

> we will have to add reasonable recommendations for every scheduler we introduce

This is certainly not the way to go. Could we not just use whatever the default is?

Comment (guybuk) [reply skorch/tests/callbacks/test_lr_scheduler.py:100]: > This is a bit of a conundrum. Do you know why some schedulers prefer batch and other epoch? Is it some derived property or more arbitrary?

Not for certain. Naturally if you wanted your lr to change smoothly you'd do it per batch rather than epoch. I'm just plugging in a scheduler as is written in its documentation.

> This is certainly not the way to go. Could we not just use whatever the default is?

A scheduler doesn't have a default. It has a step function which the user can use whenever and wherever they'd like. Setting mode `epoch` or `batch` as defaults will break backwards compatibility for anything that isn't intended to use the default mode we choose.

Since the Cyclic scheduler is a special case for which skorch hacked a solution, I suggested leaving that solution as is (so that it won't conflict with this feature or have compatibility issues) until skorch makes an update that breaks compatibility, and just get rid of that hack then.

Comment (BenjaminBossan) [reply skorch/tests/callbacks/test_lr_scheduler.py:100]: Good points. Maybe it would be worth it to actually break backwards compatibility here for the sake of getting a clean solution. In fact, we could even detect when such a breakage happens and issue a warning to the user, with instructions how to revert to the old behavior. Would that be a compromise?

Comment (thomasjpfan) [reply skorch/tests/callbacks/test_lr_scheduler.py:100]: > Would that be a compromise?

I would be happy with a deprecation warning with a suggestion of using `step_every='batch'` in the future.

File: skorch/callbacks/lr_scheduler.py (hunk 2 above)
Comment (guybuk) [reply skorch/callbacks/lr_scheduler.py]: Fixed

File: skorch/tests/callbacks/test_lr_scheduler.py (hunk 1 above)
Comment (guybuk) [reply skorch/tests/callbacks/test_lr_scheduler.py:100]: Done.

**Comment by guybuk** [comment]
Looked at the `ReduceLROnPlateau` tests, and shifted the `if` statements in `on_epoch_end` around so the tests pass. I probably need someone who knows the feature to make sure it's still ok.

File: skorch/callbacks/lr_scheduler.py [hunk 3]
```diff
@@ -158,19 +158,20 @@ def on_epoch_end(self, net, **kwargs):
...
-                    if self.lr_scheduler_.mode == 'max':
-                        score = -np.inf
-                    else:
-                        score = np.inf
+                if self.lr_scheduler_.mode == 'max':
+                    score = -np.inf
+                elif self.lr_scheduler_.mode == 'min':
+                    score = np.inf
+                elif epoch:
+                    score = net.history[-1, self.monitor]
```
Comment (BenjaminBossan) [inline skorch/callbacks/lr_scheduler.py:166]: I'm no expert on lr schedulers, so can't comment on the exact logic that should be applied here. However, what I'm seeing is that the logic might have a "hole", where score wouldn't be defined at all if none of the `if` or `elif` matches. In the logic before, score would always be defined.

Could it be that we don't actually need `elif epoch`? Previously, we accessed `net.history[-2, self.monitor]`, i.e. the second to last row in the `history`. Therefore, we had to make sure that we're not in the first epoch (at least that's my interpretation of things). Now that we're accessing `net.history[-1, self.monitor]`, we can probably never get an index error here, hence don't need the guard. Would that make sense?

Comment (guybuk) [reply skorch/callbacks/lr_scheduler.py:166]: Makes sense. Pushed a change.

File: skorch/callbacks/lr_scheduler.py [hunk 4]
```diff
@@ -66,11 +72,21 @@ def __init__(self,
...
                  **kwargs):
         self.policy = policy
         self.monitor = monitor
         self.event_name = event_name
+        self.step_every = step_every
         vars(self).update(kwargs)
+        if policy == TorchCyclicLR or policy == "TorchCyclicLR":
+            warnings.warn(
+                "The LRScheduler callback makes a step "
+                "every epoch by default from now on. To have the cyclic lr "
```
Comment (thomasjpfan) [inline skorch/callbacks/lr_scheduler.py:85]: For deprecation, we must keep the current behavior and state that the behavior will change in a Future version:

> Starting in 0.10, the LRScheduler will make a step every epoch by default. To have the cyclic lr schedule update every batch set step_every='batch'".

I am thinking of moving this logic into `initialize`:

```py
def initialize(self):
	self.step_every_ = self.step_every
    if policy == TorchCyclicLR or policy == "TorchCyclicLR" and self.step_every=='epoch':
		self.step_every_ = 'batch'
```

And then use `self.step_every_` in the other methods.

@BenjaminBossan How do you feel about this approach to deprecating the behavior? The simpler approach would be to update in `__init__`.

File: skorch/callbacks/lr_scheduler.py [hunk 5]
```diff
@@ -66,11 +72,21 @@ def __init__(self,
...
+                "scheduler update every batch, "
+                "set step_every='batch'",
+                DeprecationWarning,
```
Comment (thomasjpfan) [inline skorch/callbacks/lr_scheduler.py:88]: FutureWarning shows by default to users.

```suggestion
                FutureWarning,
```

File: skorch/callbacks/lr_scheduler.py (hunk 4 above)
Comment (guybuk) [reply skorch/callbacks/lr_scheduler.py:85]: Could you clarify the advantage of having both `self.step_every` and `self.step_every_`? It looks to me like having a second variable is unnecessary.

Once we decide whether to place this code in `initialize` or `init`, I'll make the changes.

Comment (thomasjpfan) [reply skorch/callbacks/lr_scheduler.py:85]: Traditionally the sckit-learn API does not allow for changing the `__init__` parameters during construction. But since this is a callback and not an estimator, we may not need to strictly follow this API contract.

Comment (BenjaminBossan) [reply skorch/callbacks/lr_scheduler.py:85]: Please try to stick with the sklearn convention here and don't change the `__init__` parameters. This means making the change during `initialize` (which is equivalent to `fit` in sklearn) and use the trailing underscore. The reason why we need this is because it makes it possible to use `set_params` on callbacks.

> How do you feel about this approach to deprecating the behavior?

When it comes to deprecation: I don't have a strong opinion. From my point of view, it would also be okay to change the behavior to the new behavior, raise a warning and pass a message that explains how to get back to the old behavior. But deprecation is also totally fine with me.

Regardless of how, could you please add a `# TODO: ...` comment, @guybuk, so that we don't forget to eventually change the behavior and remove the warning?

Comment (thomasjpfan) [reply skorch/callbacks/lr_scheduler.py:85]: > When it comes to deprecation: I don't have a strong opinion. From my point of view, it would also be okay to change the behavior to the new behavior, raise a warning and pass a message that explains how to get back to the old behavior. But deprecation is also totally fine with me.

In this case, I am +0.5 with moving fast and go with the behavior change.

Let's change the behavior with a FutureWarning, we can remove the warning in `0.10`. With the behavior change, we will not need to have a `self.step_every_`.

Comment (guybuk) [reply skorch/callbacks/lr_scheduler.py]: Done. Let me know if there's anything else.

Also, thanks for holding my hand throughout the process. I hope it was worth you time :smiley:

File: skorch/callbacks/lr_scheduler.py [hunk 6]
```diff
@@ -123,6 +115,15 @@ def initialize(self):
         self.policy_ = self._get_policy_cls()
         self.lr_scheduler_ = None
         self.batch_idx_ = 0
+        # TODO: Remove this warning on 0.10 release
+        if self.policy_ == TorchCyclicLR or self.policy_ == "TorchCyclicLR":
```
Comment (thomasjpfan) [inline skorch/callbacks/lr_scheduler.py:119]: Only need to warn if `step_every` is `'epoch'`.

```suggestion
        if self.policy_ == TorchCyclicLR or self.policy_ == "TorchCyclicLR" and self.step_every == 'epoch':
```

We should also have a test to make sure the warning message is raised.

File: skorch/tests/callbacks/test_lr_scheduler.py [hunk 7]
```diff
@@ -241,6 +241,32 @@ def test_lr_scheduler_record_batch_step(self, classifier_module, classifier_data
...
+        (TorchCyclicLR, {'base_lr': 123, 'max_lr': 999})
+    ])
+    def test_cyclic_lr_with_epoch_step_warning(self,
+                                               classifier_module,
+                                               classifier_data,
+                                               policy,
+                                               kwargs):
+        with pytest.warns(None) as record:
+            scheduler = LRScheduler(
+                policy, **kwargs)
```
Comment (thomasjpfan) [inline skorch/tests/callbacks/test_lr_scheduler.py:254]: We do not need to parametrize

```suggestion
                TorchCyclicLR, base_lr=123, max_lr=999)
```

File: skorch/callbacks/lr_scheduler.py [hunk 8]
```diff
@@ -116,7 +116,8 @@ def initialize(self):
         self.lr_scheduler_ = None
         self.batch_idx_ = 0
         # TODO: Remove this warning on 0.10 release
-        if self.policy_ == TorchCyclicLR or self.policy_ == "TorchCyclicLR":
+        if self.policy_ == TorchCyclicLR or self.policy_ == "TorchCyclicLR" \
+                and self.step_every == 'epoch':
```
Comment (thomasjpfan) [inline skorch/callbacks/lr_scheduler.py:120]: Nit:

```suggestion
        if (self.policy_ == TorchCyclicLR or self.policy_ == "TorchCyclicLR" 
                and self.step_every == 'epoch'):
```

File: skorch/tests/callbacks/test_lr_scheduler.py [hunk 9]
```diff
@@ -241,6 +241,32 @@ def test_lr_scheduler_record_batch_step(self, classifier_module, classifier_data
...
 
+    @pytest.mark.parametrize('policy, kwargs', [
+        (TorchCyclicLR, {'base_lr': 123, 'max_lr': 999})
+    ])
+    def test_cyclic_lr_with_epoch_step_warning(self,
+                                               classifier_module,
+                                               classifier_data,
+                                               policy,
+                                               kwargs):
+        with pytest.warns(None) as record:
```
Comment (thomasjpfan) [inline skorch/tests/callbacks/test_lr_scheduler.py:252]:
```suggestion
        msg = "The LRScheduler now makes a step every epoch by default. "
        with pytest.warns(FutureWarning, match=msg) as record:
```

After this change, we can remove the lines that come after `assert len(record) == 1`

File: CHANGES.md [hunk 10]
```diff
@@ -11,12 +11,14 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
+- Added the `step_every` argument for `LRScheduler` to set whether the scheduler step should be taken on every epoch or on every batch.
 
 ### Changed
 
 - Removed support for schedulers with a `batch_step()` method in `LRScheduler`. 
 - Raise `FutureWarning` in `CVSplit` when `random_state` is not used. Will raise an exception in a future (#620)
 - The behavior of method `net.get_params` changed to make it more consistent with sklearn: it will no longer return "learned" attributes like `module_`; therefore, functions like `sklearn.base.clone`, when called with a fitted net, will no longer return a fitted net but instead an uninitialized net; if you want a copy of a fitted net, use `copy.deepcopy` instead;`net.get_params` is used under the hood by many sklearn functions and classes, such as `GridSearchCV`, whose behavior may thus be affected by the change. (#521, #527)
+- Raise `FutureWarning` when using `CyclicLR` scheduler, because the default behavior has changed from taking a step every batch to taking a step every epoch.
```
Comment (thomasjpfan) [inline CHANGES.md:21]:
```suggestion
- Raise `FutureWarning` when using `CyclicLR` scheduler, because the default behavior has changed from taking a step every batch to taking a step every epoch. (#626)
```
